### DISCUSSION TITLE
Questions about reproducability

### DISCUSSION THREAD

**Discussion by pbsds** [photoprism/photoprism#2501] [General]
Hi, skimming through the build procedure i see that `scripts/download-nasnet.sh`, `scripts/download-facenet.sh` and `scripts/download-nsfw.sh` constrain the sha1 of the downloaded pretrained model weights. I see that the hash for nasnet has been changed at least twice.

What I'm wondering about is:

* Could a future update break old snapshots of these download scripts?
* If so, is it alright to ignore the hashes? Could the format or "interface" of the models ever change, or will future pretrained model versions always be interchangeable?
* If not, is it possible to version the artifacts at <https://dl.photoprism.app/tensorflow/>?

**Comment by lastzero** [comment]
For incompatible changes that are not just improvements, we generally change the download URL. We also keep an archive containing all dependencies tagged with date and commit ID:

- https://hub.docker.com/r/photoprism/photoprism/tags
- https://docs.photoprism.app/getting-started/faq/#why-are-you-using-docker

**Reply by pbsds** [reply]
The reason I'm asking is because i'm concidering taking a non-docker route to self host.

**Reply by lastzero** [reply]
The many dependencies and the problem of reproducibility are the reason why we use Docker:

- PhotoPrism is a highly complex application and much of its functionality depends not only on TensorFlow models, but also on FFmpeg and a number of RAW converters, with each version supporting slightly different file formats.
- This can make a big difference from a user's perspective, even if it's technically not our responsibility.
- We also can't guarantee that all dependencies will always be available in the future.

**None of this should stop you from building your own binaries (using the scripts we provide, for example), and we do our best not to break things for those who go that route. If you have a particular problem, such as an incompatible shell script, feel free to reach out.**
